docs/blog/pydantic-v2.md

```diff
@@ -0,0 +1,743 @@
+
+Partly due to the issues with the lack of strict mode,
+I've gone back and forth between using the terms "parse" and "validate" for what pydantic does.
+
+While pydantic is not simply a validation library (and I'm sure some would argue validation is not strictly what it does),
+most people use the word **"validation"**.
+
+It's time to stop fighting that, and use consistent names.
+
+The word "parse" will no longer be used except when talking about JSON parsing, see
+[model methods](#model-namespace-cleanup) above.
+
+## Changes to custom field types :neutral_face:
+
+Since the core structure of validators has changed from "a list of validators to call one after another" to
+"a tree of validators which call each other", the 
+[`__get_validators__`](https://pydantic-docs.helpmanual.io/usage/types/#classes-with-__get_validators__)
+way of defining custom field types no longer makes sense.
+
+Instead we'll look for the attribute `__pydantic_schema__` which must be a 
+pydantic-core compliant schema for validating data to this field type (the `function`
+item can be a string, if so a function of that name will be taken from the class, see `'validate'` below).
+
+Here's an example of how a custom field type could be defined:
+
+```py title="New custom field types"
+from pydantic import ValidationSchema
+
+class Foobar:
+    def __init__(self, value: str):
+        self.value = value
+
+    __pydantic_schema__: ValidationSchema = {
+        'type': 'function',
+        'mode': 'after',
+        'function': 'validate',
+        'schema': {'type': 'str'}
+    }
+
+    @classmethod
+    def validate(cls, value):
+        if 'foobar' in value:
+            return Foobar(value)
+        else:
+            raise ValueError('expected foobar')
+```
+
+What's going on here: `__pydantic_schema__` defines a schema which effectively says:
+
+> Validate input data as a string, then call the `validate` function with that string, use the returned value
+> as the final result of validation.
+
+`ValidationSchema` is just an alias to 
+[`pydantic_core.Schema`](https://github.com/samuelcolvin/pydantic-core/blob/main/pydantic_core/_types.py#L291)
+which is a type defining the schema for validation schemas.
+
+!!! note
+    pydantic-core schema has full type definitions although since the type is recursive, 
+    mypy can't provide static type analysis, pyright however can.
+
+We can probably provide one or more helper functions to make `__pydantic_schema__` easier to generate.
+
+## Other Improvements :thumbsup:
+
+Some other things which will also change, IMHO for the better:
+
+1. Recursive models with cyclic references - although recursive models were supported by pydantic V1,
+   data with cyclic references caused recursion errors, in pydantic-core cyclic references are correctly detected
+   and a validation error is raised
+2. The reason I've been so keen to get pydantic-core to compile and run with wasm is that I want all examples
```

Comment:
I can also imagine (but not sure if crazy) some frontend JS stuff talking directly to the wasm thing, without involving any Python, running the same validation done on backend on the frontend. Probably doesn't make sense, but anyway, would be cool.

Reply:
Ye, sounds good. This is what @PrettyWood and I joking refer to as "tydantic".

As soon as I get a multiyear prison sentence with access to a computer and internet access.

https://github.com/samuelcolvin/pydantic-core/issues/71

To do it properly, we would:
* build a library with a common interface to pyo3 and rust types, with a compile time flag to switch between pyo3 and rust types
* replace use of pyo3 in pydantic-core with that library
* build js/typescript bindings

Effectively it would be something like

```rust
struct GeneralDict {
  #![cfg(feature = "python")]
  inner: PyDict,
  #![cfg(not(feature = "python"))]
  inner: IndexMap<Value, Value>,
}
```
then outwardly identical interfaces.